pkg/feature/feature.go

```diff
@@ -0,0 +1,254 @@
+		if err != nil {
+			return err
+		}
+	} else if k8serrors.IsAlreadyExists(err) {
+		_, err = configMaps.Update(context.TODO(), configMap, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+	} else {
+		return err
+	}
+
+	return nil
+}
+
+func (f *Feature) addCleanup(cleanupFuncs ...Action) {
+	f.cleanups = append(f.cleanups, cleanupFuncs...)
+}
+
+type apply func(filename string) error
+
+func (f *Feature) apply(m manifest) error {
+	var applier apply
+	targetPath := m.targetPath()
+
+	if m.patch {
+		applier = func(filename string) error {
+			log.Info("patching using manifest", "feature", f.Name, "name", m.name, "path", targetPath)
+
+			return f.patchResourceFromFile(filename)
+		}
+	} else {
+		applier = func(filename string) error {
+			log.Info("applying manifest", "feature", f.Name, "name", m.name, "path", targetPath)
+
+			return f.createResourceFromFile(filename)
+		}
+	}
+
+	if err := applier(targetPath); err != nil {
+		log.Error(err, "failed to create resource", "feature", f.Name, "name", m.name, "path", targetPath)
+
+		return err
+	}
+
+	return nil
+}
+
+func (f *Feature) OwnerReference() metav1.OwnerReference {
+	return f.Spec.Tracker.ToOwnerReference()
+}
+
+// createResourceTracker instantiates FeatureTracker for a given Feature. All resources created when applying
+// it will have this object attached as an OwnerReference.
+// It's a cluster-scoped resource. Once created, there's a cleanup hook added which will be invoked on deletion, resulting
+// in removal of all owned resources which belong to this Feature.
+func (f *Feature) createResourceTracker() error {
+	tracker := &v1.FeatureTracker{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "dscinitialization.opendatahub.io/v1",
+			Kind:       "FeatureTracker",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: f.Spec.AppNamespace + "-" + convertToRFC1123Subdomain(f.Name),
+		},
+	}
+
+	foundTracker, err := f.DynamicClient.Resource(gvr.ResourceTracker).Get(context.TODO(), tracker.Name, metav1.GetOptions{})
+	if k8serrors.IsNotFound(err) {
+		unstructuredTracker, err := runtime.DefaultUnstructuredConverter.ToUnstructured(tracker)
+		if err != nil {
+			return err
+		}
+
+		u := unstructured.Unstructured{Object: unstructuredTracker}
+
+		foundTracker, err = f.DynamicClient.Resource(gvr.ResourceTracker).Create(context.TODO(), &u, metav1.CreateOptions{})
+		if err != nil {
+			return err
+		}
+	} else if err != nil {
+		return err
+	}
+
+	f.Spec.Tracker = &v1.FeatureTracker{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(foundTracker.Object, f.Spec.Tracker); err != nil {
+		return err
+	}
+
+	// Register its own cleanup
+	f.addCleanup(func(feature *Feature) error {
+		if err := f.DynamicClient.Resource(gvr.ResourceTracker).Delete(context.TODO(), f.Spec.Tracker.Name, metav1.DeleteOptions{}); err != nil && !k8serrors.IsNotFound(err) {
+			return err
+		}
+
+		return nil
+	})
+
+	return nil
+}
+
+func convertToRFC1123Subdomain(input string) string {
```

Comment:
not sure i understand why need this

Reply:
Reworked in 3f5efbd

Reply:
ok, i see. 
guess the string is just something meaningful to represent the component and servicemesh? 
if the limitation of 63 chars from k8s will be applied, just thinking to have something can be useful for ODH (default namespace opendatahub" as 10 chars) and downstream (default namespace 21 chars) 
create-service-mesh-routing-resources-for-dashboard (44 chars) will work in ODH, but not downstream

Reply:
> guess the string is just something meaningful to represent the component and servicemesh?

well... it's the name which should somewhat reveal it's reason to exist :) we might want to extend the `.Spec` to have `.AppNamespace` instead but I'm not quite sure.

> create-service-mesh-routing-resources-for-dashboard (44 chars) will work in ODH, but not downstream

it will be truncated to 63 chars, I guess that is fine at this point. WDYT?

Reply:
yeah, i saw the new test cases you added there. just need to be mindful when we set the value later on.
we should make it something like "dashboard-create-service-mesh-routing-resources", so the important part "dashboard" wont be truncated

Reply:
That's really a great point! That got me thinking that maybe `.Spec` should have extra fields such as `Component`, `Namespace` and thus be more useful. I will think a bit more and provide improvements as a separate chunk if that works for you.